org.eclipse.cdt.cross.arc.gnu/src/org/eclipse/cdt/cross/arc/gnu/common/CompilerPluginVersionsChecker.java

```diff
@@ -0,0 +1,103 @@
+package org.eclipse.cdt.cross.arc.gnu.common;
+
+import java.util.HashSet;
+
+import org.eclipse.cdt.make.core.scannerconfig.IScannerInfoCollector;
+import org.eclipse.cdt.make.core.scannerconfig.InfoContext;
+import org.eclipse.cdt.make.internal.core.scannerconfig2.PerProjectSICollector;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.ProjectScope;
+import org.eclipse.core.runtime.preferences.IEclipsePreferences;
+import org.eclipse.jface.dialogs.MessageDialog;
+import org.eclipse.swt.widgets.Display;
+import org.osgi.service.prefs.BackingStoreException;
+
+public class CompilerPluginVersionsChecker {
+
+    private final String FILENAME_WITH_PROJECT_SCOPED_PREFERENCES = "com.arc.cdt";
+    private final String UCLIBC_PLUGIN_VERSION = "arc.uclibc_plugin.version";
+    private final String ELF_PLUGIN_VERSION = "arc.elf32_plugin.version";
+
+    // These version numbers should be incremented when incompatible changes appear in these
+    // plug-ins.
+    private final int UCLIBC_PLUGIN_VERSION_NUMBER = 1;
+    private final int ELF_PLUGIN_VERSION_NUMBER = 1;
+
+    private final int UNREAL_VERSION_NUMBER = -1;
+
+    private HashSet<IProject> seenProjects = new HashSet<>();
+
+    private static CompilerPluginVersionsChecker INSTANCE = null;
+
+    private CompilerPluginVersionsChecker() {
+    }
+
+    public static CompilerPluginVersionsChecker getCompilerPluginVersionsChecker() {
+        if (INSTANCE == null) {
+            INSTANCE = new CompilerPluginVersionsChecker();
+        }
+        return INSTANCE;
+    }
+
+    void checkPluginVersions(IScannerInfoCollector collector) {
+        // Write plug-in versions to .settings/com.arc.cdt file if they are not
+        // present, check them otherwise and show warning if there is a
+        // mismatch.
+        InfoContext oC;
+        if ((collector instanceof PerProjectSICollector)) {
```

Comment:
I'd recommend making a quick return if collector is not an instance of PerProjectSICollector. This way you'd decrese indentation of code that is currently inside the if body.